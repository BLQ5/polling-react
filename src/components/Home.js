import React, { useEffect, useState } from 'react';
//import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';

import ClosedPolls from './ClosedPolls';
import LivePolls from './LivePolls';
import { pollActions } from '../actions/pollActions';

const Home = () => {
    const user = useSelector((state) => state.authentication.user);
    const dispatch = useDispatch();

    const [liveTabActive, setLiveTabActive] = useState(true);

    useEffect(() => {
        dispatch(pollActions.getAllPolls());
    }, []);

    return (
        <>
            <div className="container">
                <div className="tabs is-toggle">
                    <ul>
                        <li
                            className={liveTabActive ? 'is-active' : ''}
                            onClick={() => setLiveTabActive(!liveTabActive)}
                        >
                            <a>
                                <span className="icon">
                                    <i
                                        className="fas fa-chart-line"
                                        aria-hidden="true"
                                    ></i>
                                </span>
                                <span>Live</span>
                            </a>
                        </li>
                        <li
                            className={!liveTabActive ? 'is-active' : ''}
                            onClick={() => setLiveTabActive(!liveTabActive)}
                        >
                            <a>
                                <span className="icon">
                                    <i
                                        className="fas fa-store-slash"
                                        aria-hidden="true"
                                    ></i>
                                </span>
                                <span>Closed</span>
                            </a>
                        </li>
                    </ul>
                </div>
                {liveTabActive && <LivePolls />}
                {!liveTabActive && <ClosedPolls />}
                {/* <section className="section">
                <div className="container">
                    <div className="columns is-multiline">
                        {users.error && (
                <span className="text-danger">ERROR: {users.error}</span>
            )}
                    </div>
                </div>
            </section> */}
            </div>
        </>
    );
};

export default Home;
