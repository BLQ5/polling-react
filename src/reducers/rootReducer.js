import { combineReducers } from 'redux';
import { registration } from './registrationReducer';
import { authentication } from './authReducer';
import { alert } from './alertReducer';

export default combineReducers({
    registration,
    alert,
    authentication
});
