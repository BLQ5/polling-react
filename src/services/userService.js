import { apiUrl } from '../constants';
import { authHeader } from '../helpers/auth-header';

export const userService = {
    login,
    logout,
    register,
    getAllUsers,
    getById,
    update,
    removeUser,
    toggleAdminUser
};

//helper function
function handleResponse(response) {
    return response.text().then((text) => {
        const data = text && JSON.parse(text);
        if (!response.ok) {
            if (response.status === 401) {
                // auto logout if 401 response returned from api
                logout();
                window.location.reload(true);
            }

            const error = (data && data.message) || response.statusText;
            return Promise.reject(error);
        }

        return data;
    });
}

async function register(user) {
    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(user)
    };
    console.log('req', user);
    const response = await fetch(`${apiUrl}/users/register`, requestOptions);
    console.log('resp', response);
    return handleResponse(response);
}

async function login(username, password) {
    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ username, password })
    };

    const response = await fetch(
        `${apiUrl}/users/authenticate`,
        requestOptions
    );
    const user = await handleResponse(response);
    // store user details and jwt token in local storage to keep user logged in between page refreshes
    localStorage.setItem('user', JSON.stringify(user));
    return user;
}

function logout() {
    // remove user from local storage to log user out
    localStorage.removeItem('user');
}

async function getAllUsers() {
    const requestOptions = {
        method: 'GET',
        headers: { ...authHeader() }
    };

    const response = await fetch(`${apiUrl}/users`, requestOptions);
    return handleResponse(response);
}

async function getById(id) {
    const requestOptions = {
        method: 'GET',
        headers: { ...authHeader() }
    };

    const response = await fetch(`${apiUrl}/users/${id}`, requestOptions);
    return handleResponse(response);
}

async function update(user) {
    const requestOptions = {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' }, //...authHeader()
        body: JSON.stringify(user)
    };

    const response = await fetch(`${apiUrl}/users/${user.id}`, requestOptions);
    return handleResponse(response);
}

async function removeUser(id) {
    const requestOptions = {
        method: 'DELETE',
        headers: { ...authHeader() }
    };

    const response = await fetch(`${apiUrl}/users/${id}`, requestOptions);
    return handleResponse(response);
}

async function toggleAdminUser(id, isAdmin) {
    const requestOptions = {
        method: 'PATCH',
        headers: { ...authHeader() },
        body: JSON.stringify({ isAdmin: !isAdmin })
    };

    const response = await fetch(`${apiUrl}/users/${id}`, requestOptions);
    return handleResponse(response);
}
