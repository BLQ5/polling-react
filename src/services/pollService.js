import { apiUrl } from '../constants';
import { authHeader } from '../helpers/auth-header';
import { userService } from './userService';

export const pollService = {
    getAll,
    create,
    close,
    vote,
    getVotes,
    getAllClosedPollsWithVotes
};

//helper function
function handleResponse(response) {
    return response.text().then((text) => {
        const data = text && JSON.parse(text);
        if (!response.ok) {
            if (response.status === 401) {
                // auto logout if 401 response returned from api
                userService.logout();
                window.location.reload(true);
            }

            const error = (data && data.message) || response.statusText;
            return Promise.reject(error);
        }

        return data;
    });
}

async function getAll() {
    const requestOptions = {
        method: 'GET',
        headers: authHeader()
    };

    const response = await fetch(`${apiUrl}/polls`, requestOptions);
    return handleResponse(response);
}

async function create(poll) {
    console.log('req', poll);
    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', ...authHeader() },
        body: JSON.stringify(poll)
    };
    console.log('requestOptions', requestOptions);

    const response = await fetch(`${apiUrl}/polls/create`, requestOptions);
    console.log('res', response);
    return handleResponse(response);
}

async function close(pollId) {
    const requestOptions = {
        method: 'PATCH',
        headers: { 'Content-Type': 'application/json', ...authHeader() },
        body: JSON.stringify({ live: false })
    };

    const response = await fetch(`${apiUrl}/polls/${pollId}`, requestOptions);
    return handleResponse(response);
}

async function vote(pollId, option) {
    const requestOptions = {
        method: 'POST',
        headers: { 'Content-Type': 'application/json', ...authHeader() },
        body: JSON.stringify({ option })
    };

    const response = await fetch(
        `${apiUrl}/polls/${pollId}/vote`,
        requestOptions
    );
    return handleResponse(response);
}

async function getVotes(pollId) {
    const requestOptions = {
        method: 'GET',
        headers: { ...authHeader() }
    };

    const response = await fetch(
        `${apiUrl}/polls/${pollId}/votes`,
        requestOptions
    );
    return handleResponse(response);
}

export async function getAllClosedPollsWithVotes() {
    const requestOptions = {
        method: 'GET',
        headers: { ...authHeader() }
    };

    const response = await fetch(`${apiUrl}/polls/votes`, requestOptions);
    return handleResponse(response);
}
