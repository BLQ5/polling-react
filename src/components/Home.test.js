import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import { Router, Route } from 'react-router';
import { createMemoryHistory } from 'history';
import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import Home from './Home';
import rootReducer from '../reducers/rootReducer';
import { pollService } from '../services/pollService';
import { userService } from '../services/userService';

jest.mock('../services/pollService');
jest.mock('../services/userService');

//it.only -> isolate test
//it.skip -> skip test

describe('<Home /> for admin user', () => {
    let configureStore = null;
    //specifying initial state as normal user for this component
    let initialState = {
        authentication: {
            user: {
                id: 1,
                username: 'test-admin-user',
                isAdmin: true,
                token: 'fake-jwt-token-admin-1'
            }
        }
    };

    beforeEach(() => {
        configureStore = (initialState = {}) =>
            createStore(rootReducer, initialState, applyMiddleware(thunk));
        pollService.getAll.mockResolvedValue([
            {
                question: 'sdfsdf',
                options: ['asd', 'aw', 'cc'],
                live: true,
                id: 1
            },
            {
                question: 'eef',
                options: ['ww', 'qqq'],
                live: false,
                id: 2
            },
            {
                question: 'asdasd',
                options: ['qwedq', 'frwc', 'vvv'],
                live: true,
                id: 3
            },
            {
                question: 'qecqec',
                options: ['zcc', 'vwrvwr', 'vvv'],
                live: true,
                id: 4
            },
            {
                question: 'asasd',
                options: ['dwq', 'xsax'],
                live: true,
                id: 5
            }
        ]);
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('Compares with existing modal window snapshot when "Create Poll" button is clicked', () => {
        //since modal is already in document object and having display:none in css rule is-active
        //so this test is to match with existing snapshot to get expected behavior

        const { getByTestId } = render(
            <Provider store={configureStore(initialState)}>
                <Home />
            </Provider>
        );

        fireEvent.click(getByTestId('CreatePoll'));

        expect(getByTestId('CreatePollModal')).toMatchSnapshot();
    });

    it('Shows Live Polls tab at start', () => {
        const { getByText } = render(
            <Provider store={configureStore(initialState)}>
                <Home />
            </Provider>
        );

        expect(getByText('Currently live')).toBeInTheDocument();
    });

    it('Fires close poll API on clicking "Close"', async () => {
        pollService.close.mockResolvedValue([]);
        const { getAllByTestId } = render(
            <Provider store={configureStore(initialState)}>
                <Home />
            </Provider>
        );
        await waitFor(() => {
            expect(pollService.getAll).toHaveBeenCalledTimes(1);
        });
        fireEvent.click(getAllByTestId('CloseButton')[0]);
        expect(pollService.close).toHaveBeenCalledWith(1); //mock value of id from our mock response
        expect(pollService.close).toHaveBeenCalledTimes(1);
    });

    it('Switches to Closed tab when clicked', () => {
        pollService.getAllClosedPollsWithVotes.mockResolvedValue([
            {
                qid: 2,
                question: 'How many months till the new iPhone?',
                options: {
                    Six: 4,
                    Four: 2,
                    Two: 17
                }
            },
            {
                qid: 6,
                question: 'How did Jon Snow die?',
                options: {
                    Stabbing: 5,
                    Beheading: 2,
                    Alive: 45
                }
            }
        ]);
        const { getByTestId } = render(
            <Provider store={configureStore(initialState)}>
                <Home />
            </Provider>
        );

        fireEvent.click(getByTestId('ClosedTab'));

        expect(getByTestId('ClosedTitle')).toBeInTheDocument();
    });

    it('Switches to List of Users tab when clicked', () => {
        userService.getAllUsers.mockResolvedValue([
            {
                username: 'admin',
                isAdmin: true,
                id: 1
            },
            {
                username: '1',
                isAdmin: false,
                id: 2
            },
            {
                username: '2',
                isAdmin: false,
                id: 3
            },
            {
                username: '3',
                isAdmin: false,
                id: 4
            }
        ]);
        const { getByTestId } = render(
            <Provider store={configureStore(initialState)}>
                <Home />
            </Provider>
        );

        fireEvent.click(getByTestId('ListOfUsersTab'));

        expect(getByTestId('ListOfUsers')).toBeInTheDocument();
    });

    it('Switches to Live tab when clicked', () => {
        const { getByTestId, getByText } = render(
            <Provider store={configureStore(initialState)}>
                <Home />
            </Provider>
        );

        fireEvent.click(getByTestId('ListOfUsersTab'));
        fireEvent.click(getByTestId('LiveTab'));
        expect(getByText('Currently live')).toBeInTheDocument();

        fireEvent.click(getByTestId('ClosedTab'));
        fireEvent.click(getByTestId('LiveTab'));
        expect(getByText('Currently live')).toBeInTheDocument();
    });

    it('Renders live polls correctly on providing mock response from API', async () => {
        //mock all polls response -> five polls, four live
        //expectation -> four live polls rendered

        const history = createMemoryHistory({ initialEntries: ['/'] });
        const { getAllByTestId } = render(
            <Provider store={configureStore(initialState)}>
                <Router history={history}>
                    <Route component={Home} />
                </Router>
            </Provider>
        );
        expect(pollService.getAll).toHaveBeenCalledTimes(1);
        await waitFor(() => {
            expect(getAllByTestId('LivePollElement')).toHaveLength(4);
        });
    });
});

describe('<Home /> for normal user', () => {
    let configureStore = null;
    //specifying initial state as normal user for this component
    let initialState = {
        authentication: {
            user: {
                id: 1,
                username: 'test-user',
                isAdmin: false,
                token: 'fake-jwt-token-1'
            }
        }
    };

    beforeEach(() => {
        configureStore = (initialState = {}) =>
            createStore(rootReducer, initialState, applyMiddleware(thunk));
        pollService.getAll.mockResolvedValue([
            {
                question: 'sdfsdf',
                options: ['asd', 'aw', 'cc'],
                live: true,
                id: 1
            },
            {
                question: 'asdasd',
                options: ['qwedq', 'frwc', 'vvv'],
                live: true,
                id: 3
            },
            {
                question: 'qecqec',
                options: ['zcc', 'vwrvwr', 'vvv'],
                live: true,
                id: 4
            },
            {
                question: 'asasd',
                options: ['dwq', 'xsax'],
                live: true,
                id: 5
            }
        ]);
    });
    afterEach(() => {
        jest.clearAllMocks();
    });
    it('Shows Live Polls tab at start', () => {
        const { getByText } = render(
            <Provider store={configureStore(initialState)}>
                <Home />
            </Provider>
        );

        expect(getByText('Currently live')).toBeInTheDocument();
    });

    it('Fires vote API on clicking "Vote"', async () => {
        pollService.vote.mockResolvedValue([]);
        const { getAllByTestId } = render(
            <Provider store={configureStore(initialState)}>
                <Home />
            </Provider>
        );
        await waitFor(() => {
            expect(pollService.getAll).toHaveBeenCalledTimes(1);
        });
        fireEvent.click(getAllByTestId('radioButton')[0]); //pick first option of first poll
        fireEvent.click(getAllByTestId('VoteButton')[0]); //pick first poll
        expect(pollService.vote).toHaveBeenCalledWith(1, 'asd'); //mock value of id and option from our mock response
        expect(pollService.vote).toHaveBeenCalledTimes(1);
    });
});
