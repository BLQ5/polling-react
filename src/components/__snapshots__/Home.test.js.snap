// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> for admin user Compares with existing modal window snapshot when "Create Poll" button is clicked 1`] = `
<div
  class="modal is-active"
  data-testid="CreatePollModal"
>
  <div
    class="modal-background"
  />
  <div
    class="modal-card"
  >
    <header
      class="modal-card-head"
    >
      <p
        class="modal-card-title"
      >
        Create a poll
      </p>
      <div
        class="buttons"
      >
        <button
          class="button is-primary"
        >
          <span>
            Question
          </span>
          <span
            class="icon"
          >
            <i
              class="fas fa-plus"
            />
          </span>
        </button>
        <button
          class="button is-primary"
          disabled=""
        >
          <span>
            Question
          </span>
          <span
            class="icon"
          >
            <i
              class="fas fa-minus"
            />
          </span>
        </button>
      </div>
    </header>
    <section
      class="modal-card-body create-poll-modal-body"
    >
      <button
        class="button create-poll-modal-arrow"
        disabled=""
        style="margin-right: 1rem;"
      >
        <span
          class="icon"
        >
          <i
            class="fas fa-arrow-left"
          />
        </span>
      </button>
      <form
        class="create-poll-modal-form"
      >
        <div
          class="field"
        >
          <label
            class="label"
          >
            Question (1/1)
          </label>
          <div
            class="control"
          >
            <textarea
              class="textarea"
              name="question"
            />
          </div>
        </div>
        <div
          class="field"
        >
          <label
            class="label"
          >
            Option 1
          </label>
          <input
            class="input"
            name="option-0"
            type="text"
            value=""
          />
        </div>
        <div
          class="field"
        >
          <label
            class="label"
          >
            Option 2
          </label>
          <input
            class="input"
            name="option-1"
            type="text"
            value=""
          />
        </div>
        <div
          class="buttons is-pulled-right"
        >
          <button
            class="button is-primary"
          >
            <span>
              Option
            </span>
            <span
              class="icon"
            >
              <i
                class="fas fa-plus"
              />
            </span>
          </button>
          <button
            class="button is-primary"
            disabled=""
          >
            <span>
              Option
            </span>
            <span
              class="icon"
            >
              <i
                class="fas fa-minus"
              />
            </span>
          </button>
        </div>
      </form>
      <button
        class="button create-poll-modal-arrow"
        disabled=""
        style="margin-left: 1rem;"
      >
        <span
          class="icon"
        >
          <i
            class="fas fa-arrow-right"
          />
        </span>
      </button>
    </section>
    <footer
      class="modal-card-foot"
    >
      <button
        class="button is-primary"
      >
        <span>
          Create
        </span>
        <span
          class="icon"
        >
          <i
            class="fas fa-check"
          />
        </span>
      </button>
      <button
        class="button"
      >
        Cancel
      </button>
      <button
        class="button"
      >
        Clear
      </button>
    </footer>
  </div>
</div>
`;
